### PR TITLE
Create train-your-own-controlnet.md

### PR DESCRIPTION
**Pull request by apolinario**
V0 of the blog. Pending: 
- Images
- Review
- Real face -> Landmark Segmentation code 
- Spaces

### CODE REVIEW

File: train-your-own-controlnet.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+3. **Training the model**: Once your dataset is ready, it is time to train the model. This is the easiest part thanks to the [diffusers training script](https://github.com/huggingface/diffusers/tree/main/examples/controlnet). You'll need a GPU with at least 8GB of VRAM.
+
+[TODO: an image of our model with a condition and a generation]
+
+## 1. Planning your condition
+To plan your condition, it is useful to think of two questions: 
+1. What kind of conditioning do I want to use?
+2. Is there an already existing model that can convert 'regular' images into my condition?
+
+For our example, we thought about using a facial landmarks conditioning. Our reasoning was: 1. the general landmarks conditioned ControlNet works well. 2. Facial landmarks are a widespread enough technique, and there are multiple models that calculate facial landmarks on regular pictures 3. Could be fun to tame Stable Diffusion to follow a certain facial landmark or imitate your own facial expression.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:38]:
```suggestion
For our example, we thought about using a facial landmarks conditioning. Our reasoning was: 1. a ControlNet model conditioned on body landmarks works released with the paper, and it works well. 2. Facial landmarks are a widespread enough technique, and there are multiple models that calculate facial landmarks on regular pictures 3. Could be fun to tame Stable Diffusion to follow a certain facial landmark or imitate your own facial expression.
```

File: train-your-own-controlnet.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+[TODO: an image grid with different facial landmarks]
+
+## 2. Building your dataset
+Okay! So we decided to do a facial landmarks Stable Diffusion conditioning. So, to prepare the dataset we need: 
+- The ground truth `image`: in this case, images of faces
+- The `conditioning_image`: in this case, images where the facial landmarks are visualised
+- The `caption`: a caption that describes the images being used
+
+For this project, we decided to go with the `FaceSynthetics` dataset by Microsoft: it is a dataset that contains 100K synthetic faces. Other face research datasets with real faces such as `Celeb-A HQ`, `FFHQ` - but we decided to go with synthetic faces for this project.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:48]: Maybe link to one of our versions of the dataset in the Hub.

Comment (pcuenca) [inline train-your-own-controlnet.md:48]:
```suggestion
For this project, we decided to go with the `FaceSynthetics` dataset by Microsoft: it contains 100K synthetic faces. There are other research datasets that contain real faces such as `Celeb-A HQ`, `FFHQ`. We decided to go with synthetic faces because it was fun, and faces were available in a wide variety of different poses.
```

File: train-your-own-controlnet.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+[Image example of the `FaceSynthetics` dataset]
+
+The `FaceSynthetics` dataset sounded like a great start: it contains ground truth images of faces, and facial landmarks annotated in the iBUG 68-facial landmarks format, and a segmented image of the face.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:52]:
```suggestion
The `FaceSynthetics` dataset sounded like a great start: it contains ground truth images of faces, facial landmarks annotated in the iBUG 68-facial landmarks format, and a segmented image of the face. 
```

File: train-your-own-controlnet.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+For this project, we decided to go with the `FaceSynthetics` dataset by Microsoft: it is a dataset that contains 100K synthetic faces. Other face research datasets with real faces such as `Celeb-A HQ`, `FFHQ` - but we decided to go with synthetic faces for this project.
+
+[Image example of the `FaceSynthetics` dataset]
+
+The `FaceSynthetics` dataset sounded like a great start: it contains ground truth images of faces, and facial landmarks annotated in the iBUG 68-facial landmarks format, and a segmented image of the face. 
+
+[TODO - Image that contains a face in the dataset, a representation of the landmarks, and a representation of the final segmentation]
+
+Perfect. Right? Unfortunately, not really. Remember the second question in the "planning your condition" step - that we should have models that convert regular images to the conditioning? Turns out there was is no known model that can turn faces into the annotated landmark format of this dataset.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:56]:
```suggestion
Perfect. Right? Unfortunately, not really. Remember the second question in the "planning your condition" step - that we should have models that convert regular images to the conditioning? Turns out there is no known model that can exactly replicate the landmarks available in this dataset. In fact, the dataset was created to facilitate research on landmark extraction features.
```

File: train-your-own-controlnet.md (hunk 2 above)
Comment (pcuenca) [reply train-your-own-controlnet.md:48]: Nevermind, saw links later.

File: train-your-own-controlnet.md [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+- Use a known model that can convert any image of a face into the 68-facial landmarks format of iBUG (in our case we used the SOTA model [SPIGA](https://github.com/andresprados/SPIGA))
+- Use custom code that converts the facial landmarks into a nice illustrated mask to be used as the `conditioning_image`
+- Save that as a [Hugging Face Dataset](https://huggingface.co/docs/datasets/indexx)
+
+[Here you can find](https://huggingface.co/datasets/pcuenq/face_synthetics_spiga) the code used to convert the ground truth images from the `FaceSynthetics` dataset into the illustrated mask and save it as a Hugging Face Dataset.
+
+Now, with the ground truth `image` and the `conditioning_image` on the dataset, we are missing one step: a caption for each image. This step is highly recommended, but you can experiment with empty prompts and report back on your results. As we did not have captions for the `FaceSynthetics` dataset, we ran it through a [BLIP captioning](https://huggingface.co/docs/transformers/model_doc/blip). You can check the code used for captioning all images [here](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned)
```
Comment (pcuenca) [inline train-your-own-controlnet.md:68]:
```suggestion
Now, with the ground truth `image` and the `conditioning_image` on the dataset, we are missing one step: a caption for each image. This step is highly recommended, but you can experiment with empty prompts and report back on your results. As we did not have captions for the `FaceSynthetics` dataset, we ran it through a [BLIP captioning model](https://huggingface.co/docs/transformers/model_doc/blip). You can check the code used for captioning all images [here](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned)
```

Comment (pcuenca) [reply train-your-own-controlnet.md:68]: TODO: clarify column names between dataset and post.

File: train-your-own-controlnet.md [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+[Insert an image that contains side by side a GT Image, the conditioning image and the caption]
+
+## 3. Training the model
+With our [dataset ready](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned), it is time to train the model! Even though this was supposed to be the hardest part of the process, with the diffusers training script, it turned out to be the easiest. We used a single A100 rented for US$1.10/h on [LambdaLabs](https://lambdalabs.com). 
+
+### Our training experience
+We trained the model for 3 epochs (this means that the batch of 100K images were shown to the model 3 times) and a batch size of 4 (each step shows 4 images to the model). This turned out to be excessive and overfit (so it forgot concepts that diverge a bit of a real face, so for example "shrek" or "a cat" in the prompt would not make a shrek or a cat but rather a person, and also started to ignore styles).
```
Comment (pcuenca) [inline train-your-own-controlnet.md:78]:
```suggestion
We trained the model for 3 epochs (this means that the 100K images were shown to the model 3 times) and a batch size of 4 (each step shows 4 images to the model). This turned out to be excessive and overfit (so it forgot concepts that diverge a bit of a real face, so for example "shrek" or "a cat" in the prompt would not make a shrek or a cat but rather a person, and also started to ignore styles). 
```

File: train-your-own-controlnet.md [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+With just 1 epoch (so after the model "saw" 100K images), it already converged to following the poses and not overfit. So it worked, but... as we used the face synthetics dataset, the model ended up learning uncanny 3D-looking faces, instead of realistic faces. This makes sense given that we used a synthetic face dataset as opposed to real ones, and can be used for fun/memetic purposes. Here is the [uncannyfaces_25K](https://huggingface.co/multimodalart/uncannyfaces_25K) model. 
+
+<iframe src="https://wandb.ai/apolinario/controlnet/reports/ControlNet-Uncanny-Faces-Training--VmlldzozODcxNDY0" style="border:none;height:512px;width:100%">
+In this interactive table you can play with the dial below to go over how many training steps the model went through and how it affects the training process. At around 15K steps, it already started learning the poses. And it matured around 25K steps. Here
```
Comment (pcuenca) [inline train-your-own-controlnet.md:83]:
```suggestion
In this interactive table you can play with the dial below to go over how many training steps the model went through and how it affects the training process. At around 15K steps, it already started learning the poses. And it matured around 25K steps. 
```

File: train-your-own-controlnet.md [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+With just 1 epoch (so after the model "saw" 100K images), it already converged to following the poses and not overfit. So it worked, but... as we used the face synthetics dataset, the model ended up learning uncanny 3D-looking faces, instead of realistic faces. This makes sense given that we used a synthetic face dataset as opposed to real ones, and can be used for fun/memetic purposes. Here is the [uncannyfaces_25K](https://huggingface.co/multimodalart/uncannyfaces_25K) model. 
+
+<iframe src="https://wandb.ai/apolinario/controlnet/reports/ControlNet-Uncanny-Faces-Training--VmlldzozODcxNDY0" style="border:none;height:512px;width:100%">
+In this interactive table you can play with the dial below to go over how many training steps the model went through and how it affects the training process. At around 15K steps, it already started learning the poses. And it matured around 25K steps. Here 
+
+### How did we do the training
```
Comment (pcuenca) [inline train-your-own-controlnet.md:85]:
```suggestion
### How we did the training
```

File: train-your-own-controlnet.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+ --tracker_project_name="controlnet" \
+ --enable_xformers_memory_efficient_attention \
+ --checkpointing_steps=5000 \
+ --validation_steps=5000 \
+ --report_to wandb \
+ --push_to_hub
+```
+
+Let's break down some of the settings, and also let's go over some optimisation tips for going as low as 8GB of VRAM for training.
+- `pretrained_model_name_or_path`: The Stable Diffusion base model you would like to use
```
Comment (pcuenca) [inline train-your-own-controlnet.md:118]: Maybe (not sure) note about the original ControlNet being trained on v1.5, but we chose 2.1 base instead.

File: train-your-own-controlnet.md [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+- `output_dir`: The directory you would like your model to be saved
+- `dataset_name`: The dataset that will be used for training. In our case [Face Synthetics SPIGA with captions](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned)
+- `conditioning_image_column`: The name of the column in your dataset that contains the conditioning image (in our case `spiga_seg`)
+- `image_column`: The name of the colunn in your dataset that contains the ground truth image (in our case `image`)
+- `caption_column`: The name of the column in your dataset that contains the caption of tha image (in our case `image_caption`)
+- `resolution`: The resolution of both the conditioning and ground truth images (in our case `512x512`)
+- `learning_rate`: The learing rate. We found out that `1e-5` worked well for this examples, but you may experiment with different values ranging between `1e-4` and `2e-6`, for example.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:125]:
```suggestion
- `learning_rate`: The learing rate. We found out that `1e-5` worked well for these examples, but you may experiment with different values ranging between `1e-4` and `2e-6`, for example.
```

File: train-your-own-controlnet.md [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+- `validation_steps`: Every `x` steps the `validaton_prompt` and the `validation_image` are ran. 
+- `report_to` where to report your training to. Here we used Weights and Biases, which gave us [this nice report]().
+But reducing the `train_batch_size` from `4` to `1` may not be enough for the training to fit a small GPU, here are some additional parameters to add for each GPU VRAM size: 
+- `push_to_hub` a parameter to push the final trained model to the Hugging Face Hub.
+
+### Fitting on a 16GB VRAM GPU
+```shell 
+pip install bitsandbytes
+
+--gradient_accumulation_steps=4 \
```
Comment (pcuenca) [inline train-your-own-controlnet.md:140]:
```suggestion
--train_batch_size=1 \
--gradient_accumulation_steps=4 \
```

File: train-your-own-controlnet.md [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+--gradient_checkpointing \
+--use_8bit_adam
+```
+
```
Comment (pcuenca) [inline train-your-own-controlnet.md:144]:
```suggestion

The combination of a batch size of 1 with 4 gradient accumulation steps is equivalent to using the original batch size of 4 we used in our example. In addition, we enabled gradient checkpointing and 8-bit Adam for additional memory savings.

```

File: train-your-own-controlnet.md [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+--set_grads_to_none
+```
+
+### Fitting on a 8GB VRAM GPU
+Please follow [our guide here](https://github.com/huggingface/diffusers/tree/main/examples/controlnet#training-on-an-8-gb-gpu)
+
+## 4. Conclusion!
+This experience of training a ControlNet was a lot of fun. We succesfully trained a model that can follow real face poses - however it learned to make uncanny 3D faces instead of real 3D faces because this was the dataset it was trained on, which has its own charm and flare. 
+
+As for next steps for us - in order to create realistically looking faces, while still not using a real face dataset, one idea is running the entire `FaceSynthetics` dataset through Stable Diffusion Image2Imaage, converting the 3D-looking faces into realistically looking ones, and then trainign another ControlNet.
```
Comment (pcuenca) [inline train-your-own-controlnet.md:159]:
```suggestion
As for next steps for us - in order to create realistically looking faces, while still not using a real face dataset, one idea is running the entire `FaceSynthetics` dataset through Stable Diffusion Image2Imaage, converting the 3D-looking faces into realistically looking ones, and then training another ControlNet.
```

Comment (pcuenca) [reply train-your-own-controlnet.md:159]: Question: do we want to tease the upcoming event?

File: train-your-own-controlnet.md (hunk 11 above)
Comment (apolinario) [reply train-your-own-controlnet.md:140]: Good catch, thanks

File: train-your-own-controlnet.md (hunk 9 above)
Comment (apolinario) [reply train-your-own-controlnet.md:118]: Sounds good. I think the reason why I chose is because v2 is better with faces

File: train-your-own-controlnet.md [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+With that, we arrived to our final dataset! The [Face Synthetics SPIGA with captions](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned) contains a ground truth image, segmentation and a caption for the 100K images of the `FaceSynthetics` dataset. We are ready to train the model!
+
+[Insert an image that contains side by side a GT Image, the conditioning image and the caption]
+
+## 3. Training the model
+With our [dataset ready](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned), it is time to train the model! Even though this was supposed to be the hardest part of the process, with the diffusers training script, it turned out to be the easiest. We used a single A100 rented for US$1.10/h on [LambdaLabs](https://lambdalabs.com).
```
Comment (osanseviero) [inline train-your-own-controlnet.md:75]:
```suggestion
With our [dataset ready](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned), it is time to train the model! Even though this was supposed to be the hardest part of the process, it turned out to be the easiest with the diffusers training script. We used a single A100 rented for US$1.10/h on [LambdaLabs](https://lambdalabs.com). 
```

File: train-your-own-controlnet.md [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+
+### Our training experience
+We trained the model for 3 epochs (this means that the batch of 100K images were shown to the model 3 times) and a batch size of 4 (each step shows 4 images to the model). This turned out to be excessive and overfit (so it forgot concepts that diverge a bit of a real face, so for example "shrek" or "a cat" in the prompt would not make a shrek or a cat but rather a person, and also started to ignore styles). 
+
+With just 1 epoch (so after the model "saw" 100K images), it already converged to following the poses and not overfit. So it worked, but... as we used the face synthetics dataset, the model ended up learning uncanny 3D-looking faces, instead of realistic faces. This makes sense given that we used a synthetic face dataset as opposed to real ones, and can be used for fun/memetic purposes. Here is the [uncannyfaces_25K](https://huggingface.co/multimodalart/uncannyfaces_25K) model.
```
Comment (osanseviero) [inline train-your-own-controlnet.md:80]:
```suggestion
With just one epoch (so after the model "saw" 100K images), it already converged to following the poses and not overfit. So it worked, but... as we used the face synthetics dataset, the model learned uncanny 3D-looking faces instead of realistic faces. This makes sense given that we used a synthetic face dataset as opposed to real ones, and it can be used for fun/memetic purposes. Here is the [uncannyfaces_25K](https://huggingface.co/multimodalart/uncannyfaces_25K) model. 
```

File: train-your-own-controlnet.md [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+ --report_to wandb \
+ --push_to_hub
+```
+
+Let's break down some of the settings, and also let's go over some optimisation tips for going as low as 8GB of VRAM for training.
+- `pretrained_model_name_or_path`: The Stable Diffusion base model you would like to use
+- `output_dir`: The directory you would like your model to be saved
+- `dataset_name`: The dataset that will be used for training. In our case [Face Synthetics SPIGA with captions](https://huggingface.co/datasets/multimodalart/facesyntheticsspigacaptioned)
+- `conditioning_image_column`: The name of the column in your dataset that contains the conditioning image (in our case `spiga_seg`)
+- `image_column`: The name of the colunn in your dataset that contains the ground truth image (in our case `image`)
```
Comment (osanseviero) [inline train-your-own-controlnet.md:122]:
```suggestion
- `image_column`: The name of the column in your dataset that contains the ground truth image (in our case, `image`)
```

File: train-your-own-controlnet.md [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+- `caption_column`: The name of the column in your dataset that contains the caption of tha image (in our case `image_caption`)
```
Comment (osanseviero) [inline train-your-own-controlnet.md:123]:
```suggestion
- `caption_column`: The name of the column in your dataset that contains the caption of the image (in our case, `image_caption`)
```

File: train-your-own-controlnet.md [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+- `resolution`: The resolution of both the conditioning and ground truth images (in our case `512x512`)
+- `learning_rate`: The learing rate. We found out that `1e-5` worked well for these examples, but you may experiment with different values ranging between `1e-4` and `2e-6`, for example.
```
Comment (osanseviero) [inline train-your-own-controlnet.md:125]:
```suggestion
- `learning_rate`: The learning rate. We found out that `1e-5` worked well for these examples, but you may experiment with different values ranging between `1e-4` and `2e-6`.
```

File: train-your-own-controlnet.md [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+- `validation_image`: This is for you to take a sneak peak during training! The validation images will be ran for every amount of `validation_steps` so you can see how your training is going. Insert here a local path to an arbitrary number of conditioning images
```
Comment (osanseviero) [inline train-your-own-controlnet.md:126]:
```suggestion
- `validation_image`: This is for you to take a sneak peek during training! The validation images will be run for every amount of `validation_steps` so you can see how your training is going. Insert here a local path to an arbitrary number of conditioning images
```

File: train-your-own-controlnet.md [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+- `validation_prompt`: A prompt to be ran togehter with your validation image. Can be anything that can test if your model is training well
+- `train_batch_size`: This is the size of the training batch to fit the GPU. We can afford `4` due to having an A100, but if you have a GPU with lower VRAM we recommend bringing this value down to `1`.
+- `num_train_epochs`: Each epoch corresponds to how many times the images in the training set will be "seen" by the model. We experimented with 3 epochs, but turns out the best results required just a bit more than 1 epoch, with 3 epochs our model overfit.
```
Comment (osanseviero) [inline train-your-own-controlnet.md:129]:
```suggestion
- `num_train_epochs`: Each epoch corresponds to how many times the images in the training set will be "seen" by the model. We experimented with 3 epochs, but it turns out the best results required just a bit more than 1 epoch; with 3 epochs, our model overfits.
```

File: train-your-own-controlnet.md [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+- `checkpointing_steps`: Save an intermediary checkpoint every `x` steps (in our case `5000`). So from the total `75000` steps that the 3 epochs were used, a checkpoint was saved.
```
Comment (osanseviero) [inline train-your-own-controlnet.md:130]: "So from the total 75000 steps that the 3 epochs were used, a checkpoint was saved." was a bit hard to parse

File: train-your-own-controlnet.md [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+- `train_batch_size`: This is the size of the training batch to fit the GPU. We can afford `4` due to having an A100, but if you have a GPU with lower VRAM we recommend bringing this value down to `1`.
+- `num_train_epochs`: Each epoch corresponds to how many times the images in the training set will be "seen" by the model. We experimented with 3 epochs, but turns out the best results required just a bit more than 1 epoch, with 3 epochs our model overfit.
+- `checkpointing_steps`: Save an intermediary checkpoint every `x` steps (in our case `5000`). So from the total `75000` steps that the 3 epochs were used, a checkpoint was saved. 
+- `validation_steps`: Every `x` steps the `validaton_prompt` and the `validation_image` are ran. 
+- `report_to` where to report your training to. Here we used Weights and Biases, which gave us [this nice report]().
+But reducing the `train_batch_size` from `4` to `1` may not be enough for the training to fit a small GPU, here are some additional parameters to add for each GPU VRAM size: 
+- `push_to_hub` a parameter to push the final trained model to the Hugging Face Hub.
```
Comment (osanseviero) [inline train-your-own-controlnet.md:134]:
```suggestion
- `push_to_hub`: a parameter to push the final trained model to the Hugging Face Hub.
```

File: train-your-own-controlnet.md (hunk 13 above)
Comment (osanseviero) [reply train-your-own-controlnet.md:162]: I think it would be good!

**Review by osanseviero — COMMENTED** [review]
Very cool blog post! Just small suggestions

File: train-your-own-controlnet.md [hunk 23]
```diff
@@ -0,0 +1,159 @@
...
+
+[TODO: Embed Spaces for running the model] 
+
+## Getting started with training your ControlNet for Stable Diffusion
+Training your own ControlNet training requires 3 steps: 
+1. **Planning your condition**: ControlNet is flexible enough to tame Stable Diffusion towards many tasks. The pre-trained models showcase a wide-range of conditions, and the community has built others, such as conditioning on [pixelated color pallets](https://huggingface.co/thibaud/controlnet-sd21-color-diffusers).
+
+2. **Building your dataset**: Once a condition is decided, it is time to build your dataset. For that, you can either construct a dataset from scratch, or use a sub-set of an existing dataset. You need three columns on your dataset to train the model: a ground truth `image`, a `conditioning_image` and a `prompt`. 
+
+3. **Training the model**: Once your dataset is ready, it is time to train the model. This is the easiest part thanks to the [diffusers training script](https://github.com/huggingface/diffusers/tree/main/examples/controlnet). You'll need a GPU with at least 8GB of VRAM.
```
[the remark this answers is not in the document]
Comment (apolinario) [reply train-your-own-controlnet.md:34]: IMO it's fine if we don't mention that as this is a diffusers end-to-end tutorial

File: train-your-own-controlnet.md [hunk 24]
```diff
@@ -0,0 +1,162 @@
+---
```
[the remark this answers is not in the document]
Comment (apolinario) [reply train-your-own-controlnet.md]: Done
